src/compute/cast/binary_to.rs

```diff
@@ -31,26 +33,35 @@ pub fn binary_large_to_binary(
 }
 
 /// Casts a [`BinaryArray`] to a [`PrimitiveArray`], making any uncastable value a Null.
-pub fn binary_to_primitive<O: Offset, T>(from: &BinaryArray<O>, to: &DataType) -> PrimitiveArray<T>
+pub fn binary_to_primitive<O: Offset, T>(
+    from: &BinaryArray<O>,
+    to: &DataType,
+    options: CastOptions,
+) -> PrimitiveArray<T>
 where
     T: NativeType + lexical_core::FromLexical,
 {
-    let iter = from
-        .iter()
-        .map(|x| x.and_then::<T, _>(|x| lexical_core::parse(x).ok()));
+    let parse_fn = if !options.partial {
```

Comment:
I think we need to put the whole `PrimitiveArray::<T>::from_trusted_len_iter(iter).to(to.clone())` inside the if otherwise the compiler can't inline the function call as it branches depending on the parameter.